### PR TITLE
Allows setting product user groups

### PR DESCRIPTION
**Pull request by A-Guldborg**
This allows setting the product user groups when adding and updating products, including seeing products user groups.

This is necessary in order to manage products (enabling some products only to baristas etc.)

Front-end product management pull request for Shifty:
https://github.com/AnalogIO/shifty-webapp/pull/20

### CODE REVIEW

File: coffeecard/CoffeeCard.Models/DataTransferObjects/v2/Product/AddProductRequest.cs [hunk 1]
```diff
@@ -63,6 +63,11 @@ public class AddProductRequest
         [DefaultValue(true)]
         public bool Visible { get; set; } = true;
 
+        /// <summary>
+        /// Gets or sets the user groups that can access the product.
+        /// </summary>
+        /// <value> Product User Groups </value>
+        /// <example> Customer </example>
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.Models/DataTransferObjects/v2/Product/AddProductRequest.cs:70]: This is a list of UserGroups, thus the example should be an array. Please also remember to update the example block above the class declaration

File: coffeecard/CoffeeCard.Models/DataTransferObjects/v2/Product/ChangedProductResponse.cs [hunk 2]
```diff
@@ -51,5 +52,13 @@ public class ChangedProductResponse
...
         [Required]
         public bool Visible { get; set; }
+
+        /// <summary>
+        /// Gets or sets the user groups that can access the product.
+        /// </summary>
+        /// <value> Product User Groups </value>
+        /// <example> Customer </example>
+        [Required]
+        public IEnumerable<UserGroup> AllowedUserGroups { get; set; }
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.Models/DataTransferObjects/v2/Product/ChangedProductResponse.cs:62]: Same comment as added in `AddProductRequest`

File: coffeecard/CoffeeCard.Models/DataTransferObjects/v2/Product/UpdateProductRequest.cs [hunk 3]
```diff
@@ -69,5 +71,13 @@ public class UpdateProductRequest
...
         [DefaultValue(true)]
         public bool Visible { get; set; } = true;
+
+        /// <summary>
+        /// Gets or sets the user groups that can access the product.
+        /// </summary>
+        /// <value> Product User Groups </value>
+        /// <example> Customer </example>
+        [Required]
+        public IEnumerable<UserGroup> AllowedUserGroups { get; set; }
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.Models/DataTransferObjects/v2/Product/UpdateProductRequest.cs:81]: Same comment as in `AddProductRequest`

File: coffeecard/CoffeeCard.Models/DataTransferObjects/v2/Products/ProductResponse.cs [hunk 4]
```diff
@@ -64,5 +66,13 @@ public class ProductResponse
...
         [Required]
         public bool IsPerk { get; set; }
+
+        /// <summary>
+        /// Decides the user groups that can access the product.
+        /// </summary>
+        /// <value> Product User Groups </value>
+        /// <example> Customer </example>
+        [Required]
+        public IEnumerable<UserGroup> AllowedUserGroups { get; set; }
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.Models/DataTransferObjects/v2/Products/ProductResponse.cs:76]: Same comment as in AddProductRequest

File: coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs [hunk 5]
```diff
@@ -119,13 +119,28 @@ public async Task<ChangedProductResponse> UpdateProduct(UpdateProductRequest cha
 
             await _context.SaveChangesAsync();
 
+            var existingUserGroups = _context.ProductUserGroups.Where(e => e.ProductId == changedProduct.Id);
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs:122]: The `GetProductAsync` method already returns a Product where its ProductUserGroups are included, thus there is no reason to call the DB to find the ProductUserGroup. Instead you can just manipulate the property directly as done for the other properties above.

Comment (jonasanker) [reply coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs:122]: Please also write a unit test for the above scenario, thanks!

**Review by jonasanker — CHANGES_REQUESTED** [review]
Few changes requested :)

**Comment by jonasanker** [comment]
Be aware that I'm cleaning up parts of the Products endpoints in #224

**Review by jonasanker — CHANGES_REQUESTED** [review]

File: coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs [hunk 6]
```diff
@@ -119,13 +119,28 @@ public async Task<ChangedProductResponse> UpdateProduct(UpdateProductRequest cha
...
+
+            _context.RemoveRange(existingUserGroups);
+
+            var newProductUserGroups = changedProduct.AllowedUserGroups.Select(userGroup => new ProductUserGroup
+            {
+                ProductId = product.Id,
+                UserGroup = userGroup
+            }).ToList();
+
+            _context.AddRange(newProductUserGroups);
```
[the remark this answers is not in the document]
Comment (A-Guldborg) [reply coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs]: Ah, makes sense. I did not properly read the previous comment I suppose :P 

I have added it, both for these lines as well as 84-93 (though it is not as clean in 84-93 due to needing an instantiated product before I can add them, but without the redundant DB call). Please critically read my test to verify that it actually tests no functionality breaks with this change!

File: coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs [hunk 7]
```diff
@@ -81,41 +81,43 @@ public async Task<ChangedProductResponse> AddProduct(AddProductRequest newProduc
...
+            product.ProductUserGroup = newProduct.AllowedUserGroups.Select(userGroup => new ProductUserGroup
             {
                 ProductId = product.Id,
                 UserGroup = userGroup
             }).ToList();
 
-            _context.ProductUserGroups.AddRange(productUserGroups);
-
-
+            _context.Products.Add(product);
```
[the remark this answers is not in the document]
Comment (A-Guldborg) [reply coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs:90]: I am uncertain what you mean with "re-add the Products":
This product is not added to the context yet. It is created in line 74, added user groups to the product on line 84 and then added to the context on line 90 before being saved.

My new test, `AddProduct adds only selected user groups`, fails with error message `CoffeeCard.Common.Errors.EntityNotFoundException : No product was found by Product Id: 1 ` if I remove this line where I add it to the context. Could it be that since the changes now also involves the `AddProduct` method that you confused this with the `UpdateProduct` method?

File: coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs [hunk 8]
```diff
@@ -81,41 +81,43 @@ public async Task<ChangedProductResponse> AddProduct(AddProductRequest newProduc
                 Visible = newProduct.Visible
             };
 
-            _context.Products.Add(product);
-            await _context.SaveChangesAsync();
-
-            var productUserGroups = newProduct.AllowedUserGroups.Select(userGroup => new ProductUserGroup
+            product.ProductUserGroup = newProduct.AllowedUserGroups.Select(userGroup => new ProductUserGroup
```
[the remark this answers is not in the document]
Comment (A-Guldborg) [reply coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs:84]: I thought it was not possible but I guess I didn't play around enough with it. It is fixed in the new commit :)

The ProductUserGroup type has 3 properties: Product, ProductId and UserGroup. I assumed I could not just add the UserGroup without a reference to the product, but just "casting" the `UserGroup` to `ProductUserGroup` I still got the correct result.

File: coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs [hunk 9]
```diff
@@ -81,41 +81,43 @@ public async Task<ChangedProductResponse> AddProduct(AddProductRequest newProduc
...
+                Visible = product.Visible,
+                AllowedUserGroups = newProduct.AllowedUserGroups
             };
 
             return result;
         }
 
         public async Task<ChangedProductResponse> UpdateProduct(UpdateProductRequest changedProduct)
         {
+            var newProductUserGroups = changedProduct.AllowedUserGroups.Select(userGroup => new ProductUserGroup
```
[the remark this answers is not in the document]
Comment (A-Guldborg) [reply coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs:108]: True, done in the new commit :)

File: coffeecard/CoffeeCard.Tests.Unit/Services/v2/ProductServiceTest.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+using CoffeeCard.Models.DataTransferObjects.v2.Product;
+using CoffeeCard.Models.Entities;
+using Microsoft.EntityFrameworkCore;
+using Xunit;
+
+namespace CoffeeCard.Tests.Unit.Services.v2
+{
+    public class ProductServiceTest
+    {
+        [Fact(DisplayName = "UpdateProduct removes ommitted user groups and only adds selected user groups")]
```
[the remark this answers is not in the document]
Comment (A-Guldborg) [reply coffeecard/CoffeeCard.Tests.Unit/Services/v2/ProductServiceTest.cs:17]: Please elaborate when you say "an existing product".

If you see lines 46-56 you will see that I add `Barista` and `Manager` as the user groups on the database and save it to the context before creating the product service on line 60. Then I update the user groups to be `[Customer, Board]` and in line 78 I retrieve the product through the service and verify it has exactly elements `Customer` and `Board` as user groups. 

So in short, I believe this test already tests this behaviour. Does that make sense?

File: coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jonasanker) [reply coffeecard/CoffeeCard.Library/Services/v2/ProductService.cs:88]: You are right. I messed up the lines. Looks good now! 😊

**Review by jonasanker — APPROVED** [review]
Great work. Approved ✅
